### ISSUE TITLE
Support synthetic source for all field types

### ISSUE THREAD

**Issue by salvatore-campagna** [elastic/elasticsearch#106460]
### Description

Some field types are still missing support for synthetic source. We would like to implement the missing functionality adopting a strategy that allows us to use a stored field behind the scenes. The idea is to use the stored field later on to support reconstructing the original field value instead of reconstructing it from doc values.

The ones we expect to be used more frequently for logging and that we should prioritise when it comes to implementing support are:
- [x] binary #107549
- [x] object (`enabled: false`) #108051
- [x] nested #108818
- [x] long_range #107081
- [x] double_range #107081
- [x] date_range #107081
- [x] ip_range #107081
- [x] murmur3
- [x] annotated-text #107735

For fields that are less frequently used we will implement a less storage efficient fallback implementation of synthetic source. This implementation will cover all fields that are not currently supported and unblock their usage in logs index mode. Based on usage of such fields, more efficient implementation can be added.
- [ ] join
- [ ] completion
- [x] search_as_you_type #108222, #108878
- [x] token_count #109044
- [ ] sparse_vector
- [ ] rank_feature
- [ ] rank_features
- [x] geo_shape #108881
- [ ] point #109312
- [ ] shape #109312
- [ ] percolator

We would like to enable this feature by default if synthetic source is enable. This will enable usage of these field types for both logs and TSDB index mode.
In general support should be provided by making the `stored` property default to `true` when synthetic source is enabled. This implies we need to add the `stored` property even for fields that are missing it.

Once support is implemented please check the box next to the corresponding type and link the PR next to it.

**Comment by salvatore-campagna** [comment]
After data shared by @ruflin about usage of field types I am changing the list of fields to be supported adding `annotated-text` to the list of fields to support. This is basically a text field which we expect to deal with the same way we deal with other text fields, by having a hidden stored fields used only if it does not have the corresponding `keyword` field.

**Comment by lkts** [comment]
Recommended usage of `murmur3` field is as a multi field to achieve a speedup in cardinality queries. As a multi field it won't impact synthetic source generation, source will be generated using parent field. Hash values are always stored as doc values and therefore are available no matter what is in source.

In theory it is possible to create a mapping with plain `murmur3` field but it is not clear what use case that would achieve.

In summary, `murmur3` does not require any additional logic.
